Derek/20230627 activate pro accounts with emails address

1orZero opened the pull request:
- Add the **email** column to **Profiles** table
 - Update the **Stripe webhook** to handle updating the user to the Pro plan **by email address**.
 - Show the **Enjoy trial message** only when the user is in trial. Add the necessary handling for it.
 - Update the **handle_new_user** trigger function for the **new email columns**.
 - Update **REFERRAL_TRIAL_DAYS** to **NEXT_PUBLIC_REFERRAL_TRIAL_DAYS** to enable the frontend to utilize this environment variable.
   - I have added **NEXT_PUBLIC_REFERRAL_TRIAL_DAYS** env variable in **Vercel**.
---
Task left for @thejackwu 
- [x] Remove **REFERRAL_TRIAL_DAYS** env variable from Vercel once deployed to production.
- [x] Run Supabase Migration in production

1orZero committed:
add email column & handle webhook

1orZero committed:
Update customer.subscription.updated event, remove extend pro account date

1orZero committed:
show Enjoy trial message only when they are in trial

1orZero committed:
update handle_new_user trigger function for the new email columns

thejackwu commented on pages/api/webhooks/stripe.ts:
```diff
@@ -42,18 +42,44 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+        throw new Error('plan code from Stripe webhook');
+      }
+
+      if (!userId && !email) {
```
I would check only email here. Because if someone purchases through the Stripe link directly, they won't have userID.

1orZero replied:
DONE

thejackwu commented:
I assume we are still missing a script that can populate the existing profiles table with emails?

1orZero commented:
> I assume we are still missing a script that can populate the existing profiles table with emails? @1orZero 

Yes, I thought you have the script setup so i will leave it up to you. I can also do that tomorrow as well . Its should be easy.

1orZero committed:
Add handles to plan_giving_weeks

1orZero committed:
Remove console.log

1orZero commented:
I have added support for `plan_giving_weeks` metadata, which takes priority over `plan_code`.

The program will now update the `pro_plan_expiration_date` based on the given number of weeks from the metadata of Stripe product.

1orZero committed:
Update the webhook to check only email field

thejackwu commented on pages/api/webhooks/stripe.ts:
```diff
@@ -40,35 +40,60 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+      const proPlanExpirationDate = (() => {
+        // Takes plan_giving_weeks priority over plan_code
+        if (planGivingWeeks && typeof planGivingWeeks === 'string') {
+          const currentDate = dayjs().utc().toDate();
+          return dayjs(currentDate).add(+planGivingWeeks, 'week').toDate();
+        } else if (planCode === ONE_TIME_PRO_PLAN_FOR_1_MONTH) {
+          // Only store expiration for one month plan
+          const currentDate = dayjs().utc().toDate();
+          return dayjs(currentDate).add(1, 'month').toDate();
```
The "Since" date, can we use the Stripe payment date instead? Because there are webhooks that is being retried at Stripe which is from weeks ago @1orZero

thejackwu replied:
Also if user has a expiration date already, can we add the planGIvingWeeks on top of that instead? Since user will purchase more later on

1orZero replied:
DONE

1orZero committed:
Handle the current date by using session created date from stripe

1orZero committed:
userProfileQuery handles by email

1orZero committed:
Add ability to extend pro expiration date if planGivingWeeks provided

1orZero committed:
add resend email service on stripe webhook error

1orZero commented:
- Add Resend Email service address 
  - Add env variable
    - RESEND_EMAIL_API_KEY=KEY
    - RESEND_EMAIL_RECEIVER=EMAIL
  - Task left for @thejackwu 
     - [ ] Add a verified domain for the from email to Resend or use their domain
     - [ ] Add the env variable in Vercel

1orZero commented on utils/server/resend.ts:
```diff
@@ -0,0 +1,68 @@
+
+export async function sendReport(subject = '', html = '') {
+  await resend.emails.send({
+    // TODO: need to change this to a verified email domain
```
need to update here as well

1orZero commented on utils/server/resend.ts:
```diff
@@ -0,0 +1,68 @@
+export async function sendReport(subject = '', html = '') {
+  await resend.emails.send({
+    // TODO: need to change this to a verified email domain
+    from: 'report@resend.dev',
```
need to update here as well @thejackwu

thejackwu committed:
Update from email address

thejackwu committed:
Fix vulnerable packages

thejackwu commented on pages/api/webhooks/stripe.ts:
```diff
@@ -40,41 +47,102 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+    // Monthly Pro Plan Subscription recurring payment
     if (event.type === 'customer.subscription.updated') {
       const session = event.data.object as Stripe.Subscription;
       stripeSubscriptionId = session.id;
 
       if (!stripeSubscriptionId) {
-        console.error(session);
+        await sendReportForStripeWebhookError(
+          'Subscription id not found from Stripe webhook',
+          event,
+        );
         throw new Error('Subscription id not found from Stripe webhook');
```
Can we also use customer email to look for them, if the stripe subscriptionID is not present?

thejackwu commented on pages/api/webhooks/stripe.ts:
```diff
@@ -96,17 +164,21 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
 
       if (!stripeSubscriptionId) {
-        console.error(session);
+        await sendReportForStripeWebhookError(
+          'Subscription id not found from Stripe webhook',
+          event,
+        );
```
Same as above

1orZero committed:
Detect voice input and show indicator

1orZero committed:
Refactor stripe webhook, update error handling

1orZero commented:
I refactored the Stripe webhook and updated the error handling.
Most cases should not throw errors in the webhooks. Instead, they send the error via Resend.

thejackwu committed:
Add console.log for debuggin

thejackwu committed:
Update email sending mechanism to use await instead of Promise

thejackwu committed:
Remove console.logs
